File: Employee_Absenteeism .py

# coding: utf-8

# In[1]:


# Importing Libraries
import pandas as pd
from pandas import Timestamp
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import chi2_contingency
from fancyimpute import KNN 
import os
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from scipy import stats
import statsmodels.api as sm
import seaborn as sns
import matplotlib.gridspec as gridspec 
get_ipython().run_line_magic('matplotlib', 'inline')
from ggplot import *


# In[2]:


#set working directory
os.chdir("S:/analytics/Projects/Employee Absenteeism python")


# In[3]:


os.getcwd()


# In[4]:


#loading the data 
data = pd.read_excel("Absenteeism_at_work_Project.xls")


# In[5]:


data


# In[10]:


data.dtypes


# In[8]:


data.shape


# # data pre processing

# In[9]:


#Exploratory Data Analysis
data['Reason for absence']=data['Reason for absence'].astype(object)
data['Month of absence']=data['Month of absence'].astype(object)
data['Day of the week']=data['Day of the week'].astype(object)
data['Seasons']=data['Seasons'].astype(object)
data['Service time']=data['Service time'].astype(object)
data['Hit target']=data['Hit target'].astype(object)
data['Disciplinary failure']=data['Disciplinary failure'].astype(object)
data['Education']=data['Education'].astype(object)
data['Son']=data['Son'].astype(object)
data['Social drinker']=data['Social drinker'].astype(object)
data['Social smoker']=data['Social smoker'].astype(object)
data['Pet']=data['Pet'].astype(object)


# In[16]:


#finding the number of missing values
data.isnull().sum().sum()


# In[17]:


# Categorising into continuous and categorical
continuous_vars = ['Distance from Residence to Work', 'Service time', 'Age', 'Work load Average/day ', 'Transportation expense',
       'Hit target', 'Weight', 'Height', 'Body mass index', 'Absenteeism time in hours']

categorical_vars = ['ID','Reason for absence','Month of absence','Day of the week',
                     'Seasons','Disciplinary failure', 'Education', 'Social drinker',
                     'Social smoker', 'Pet', 'Son']


# # missing value analysis

# In[19]:


#Creating dataframe with missing values present in each variable
missing_val = pd.DataFrame(data.isnull().sum()).reset_index()


# In[20]:


missing_val


# In[21]:


#Renaming variables of missing_val dataframe
missing_val = missing_val.rename(columns = {'index': 'Variables', 0: 'Missing_percentage'})


# In[22]:


#Calculating percentage missing value
missing_val['Missing_percentage'] = (missing_val['Missing_percentage']/len(data))*100


# In[25]:


missing_val


# In[24]:


# Sorting missing_val in Descending order
missing_val = missing_val.sort_values('Missing_percentage', ascending = False).reset_index(drop = True)


# # Imputing missing values

# In[26]:


data['Body mass index'].iloc[12]


# In[27]:


#creating missing value
data['Body mass index'].iloc[12] = np.nan


# In[28]:


data['Body mass index'].iloc[12]


# In[29]:


# Checking for "Body mass index" column
# Actual value = 23
# Mean = 26.68
# Median = 25
# KNN = 23.20


# In[30]:


# Imputing with mean
#data['Body mass index'] = data['Body mass index'].fillna(data['Body mass index'].mean())
# data['Body mass index'].iloc[12]

# Imputing with median
#data['Body mass index'] = data['Body mass index'].fillna(data['Body mass index'].median())
# data['Body mass index'].iloc[12]


# In[31]:


#Apply KNN imputation algorithm
data = pd.DataFrame(KNN(k = 5).complete(data), columns = data.columns)


# In[32]:


data['Body mass index'].iloc[12]


# In[33]:


#confirming for missing values
data.isnull().sum()


# # outlier analysis

# In[41]:


# Ploting BoxPlot of continuous variables
plt.boxplot([data['Transportation expense']])
plt.xlabel('Transportation expense')
plt.title("BoxPlot of Variables 'Transportation expense' ")
plt.ylabel('Values')


# In[40]:


# Ploting BoxPlot of continuous variables
plt.boxplot(data['Height'])
plt.xlabel('Height')
plt.title("BoxPlot of Variables 'Height' ")
plt.ylabel('Values')


# In[42]:


plt.boxplot(data['Work load Average/day '])
plt.xlabel("Work load Average/day ")
plt.title("BoxPlot of Variable 'Work load Average/day '")
plt.ylabel('Values')


# In[43]:


plt.boxplot([ data['Distance from Residence to Work'], data['Service time'], data['Age'], data['Hit target'], data['Weight'], data['Body mass index']])
plt.xlabel(['1. Distance from Residence to Work', '2. Service time', '3. Age', '4. Hit target', '5. Weight', '6. Body mass index'])
plt.title("BoxPlot of rest of the Variables")
plt.ylabel('Values')


# In[44]:


# From the above boxplot we can clearly see that in variables 'Distance from Residence to Work', 'Weight' and 'Body mass index'
# there is no outlier


# In[45]:


# list of variables which doesn't have outlier
neglect = ['Distance from Residence to Work', 'Weight', 'Body mass index']


# In[50]:


#detect and replace outliers with NAs
for i in continuous_vars:
    #getting 75 & 25percentile of variable 'i'
    q75, q25 = np.percentile(data[i], [75,25])
    #calculating inter quartile range
    iqr = q75 - q25
    
    #calculating upper fence and lower fence
    minimum = q25 - (iqr*1.5)
    maximum = q75 + (iqr*1.5)
    
    


# In[52]:


maximum


# In[55]:


# Replacing all the outliers value to NA
data.loc[data[i]< minimum,i] = np.nan
data.loc[data[i]> maximum,i] = np.nan


# In[57]:


#checking for NAs
data.isnull().sum().sum()


# In[58]:


# Imputing missing values with KNN
data = pd.DataFrame(KNN(k = 5).complete(data), columns = data.columns)


# In[59]:


#checking for nas after knn imputation
data.isnull().sum().sum()


# # Feature selection

# In[60]:


from scipy.stats import chi2_contingency
import seaborn as sns


# In[61]:


#correlation analysis for continuous_vars
#correlation plot
df_corr = data.loc[:, continuous_vars]


# In[62]:


df_corr.shape


# In[75]:


#set the width and height of correlation plot
f, ax = plt.subplots(figsize=(7,5))

#generate correlation matrix
corr = df_corr.corr()

#plot with seaborn library
sns.heatmap(corr, mask = np.zeros_like(corr, dtype=np.bool), cmap = sns.diverging_palette(220, 10, as_cmap = True),
           square = True, ax=ax)
plt.savefig('correlation.png')


# In[77]:


#Chisquare test of independence
#Save categorical variables
catorical_vars = ["Reason for absence", "Month of absence", "Day of the week", "Seasons", "Service time", "Disciplinary failure", "Education", "Son", "Social drinker","Social smoker","Pet"]


# In[78]:


continuous_vars , categorical_vars


# In[81]:


#loop for chi square values
for i in categorical_vars:
    print(i)
    chi2, p, dof, ex = chi2_contingency(pd.crosstab(data['Absenteeism time in hours'], data[i]))
    print(p)


# In[87]:


#as we can see weight and body mass index are highly correlated with each other
# Droping the variables which has redundant information
to_drop = ['Weight']
data = data.drop(to_drop, axis = 1)
data = data.drop([ 'ID', 'Education', 'Social smoker', 'Pet'], axis=1)


# In[90]:


# Updating the Continuous Variables and Categorical Variables after droping some variables
continuous_vars = [i for i in continuous_vars if i not in to_drop]
categorical_vars = [i for i in categorical_vars if i not in to_drop]


# In[88]:


clean_data = data.copy()


# # Feature scaling

# In[98]:


##normality check
for i in continuous_vars:
    if i == 'Absenteeism time in hours':
        continue
    sns.distplot(data[i],bins = 'auto')
    plt.title("Checking Distribution for Variable "+str(i))
    plt.ylabel("Density")
    plt.show()


# In[108]:


#as we can see data is not normally distributed , we will go for normalization 
for i in continuous_vars:
    print(i)
    data[i] = (data[i] - data[i].min())/(data[i].max()-data[i].min())


# In[109]:


data


# # data sampling

# In[113]:


from sklearn.cross_validation import train_test_split
#Divide data into train and test
train, test = train_test_split(data, test_size=0.20, random_state=42)


# # machine learning models

# In[115]:


from sklearn.tree import DecisionTreeRegressor
#Decision tree for regression
fit_DT = DecisionTreeRegressor(max_depth=2).fit(train.iloc[:,0:9], train.iloc[:,9])


# In[116]:


fit_DT


# In[117]:


#checking for any missing valuses that has leeked in
np.where(data.values >= np.finfo(np.float64).max)


# In[118]:


np.isnan(data.values.any())


# In[119]:


test = test.fillna(train.mean())


# In[120]:


#Decision tree for regression
fit_DT = DecisionTreeRegressor(max_depth=2).fit(train.iloc[:,0:15], train.iloc[:,15])


# In[121]:


data.shape


# In[122]:


#Apply model on test data
predictions_DT = fit_DT.predict(test.iloc[:,0:15])


# In[123]:


predictions_DT


# In[124]:


def rmse(predictions, targets):
    return np.sqrt(((predictions - targets) ** 2).mean())


# In[125]:


rmse(test.iloc[:,15], predictions_DT)


# In[126]:


#rmse using DT = 0.16972039562573937


# In[127]:


#Divide data into train and test
X = data.values[:, 0:15]
Y = data.values[:,15]

X_train, X_test, y_train, y_test = train_test_split( X, Y, test_size = 0.2)


# In[128]:


X_test.shape


# In[133]:


from sklearn.ensemble import RandomForestRegressor


# In[134]:


# Building model on top of training dataset
fit_RF = RandomForestRegressor(n_estimators = 500).fit(X_train,y_train)


# In[135]:


fit_RF


# In[146]:


# Calculating RMSE for training data to check for over fitting
RF_predictions_train = fit_RF.predict(X_train)
rmse_for_train = np.sqrt(mean_squared_error(y_train,RF_predictions_train))


# In[145]:


# Calculating RMSE for test data to check accuracy
RF_predictions_test = fit_RF.predict(X_test)
rmse_for_test =np.sqrt(mean_squared_error(y_test,RF_predictions_test))


# In[143]:


rmse_for_train


# In[139]:


rmse_for_test


# In[140]:


print("Root Mean Squared Error For Training data = "+str(rmse_for_train))
print("Root Mean Squared Error For Test data = "+str(rmse_for_test))


# # Linear regression

# In[147]:


# Importing libraries for Linear Regression
from sklearn.linear_model import LinearRegression

# Building model on top of training dataset
fit_LR = LinearRegression().fit(X_train , y_train)

# Calculating RMSE for training data to check for over fitting
LR_pred_train = fit_LR.predict(X_train)
rmse_for_train = np.sqrt(mean_squared_error(y_train,LR_pred_train))

# Calculating RMSE for test data to check accuracy
LR_pred_test = fit_LR.predict(X_test)
rmse_for_test =np.sqrt(mean_squared_error(y_test,LR_pred_test))

print("Root Mean Squared Error For Training data = "+str(rmse_for_train))
print("Root Mean Squared Error For Test data = "+str(rmse_for_test))


# # data visualization

# In[148]:


#Visualising Important components

ggplot(clean_data, aes(x='Month of absence', y='Absenteeism time in hours')) +    geom_bar(fill= "SlateBlue") +    scale_color_brewer(type='diverging', palette=4) +    xlab("Month") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[152]:


ggplot(clean_data, aes(x='Age', y='Absenteeism time in hours')) +    geom_bar(fill= "Gray") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Age") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[153]:


ggplot(clean_data, aes(x='Social drinker', y='Absenteeism time in hours')) +    geom_bar(fill= "Magenta") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Social drinker") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[154]:


ggplot(clean_data, aes(x='Body mass index', y='Absenteeism time in hours')) +    geom_bar(fill= "Yellow") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Body mass index") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[175]:


sns.stripplot(x="Reason for absence", y="Absenteeism time in hours", data=data,size = 10);
plt.savefig('Reason for absence.png')


# In[158]:


sns.stripplot(x="Month of absence", y="Absenteeism time in hours", data=data, size = 5);
plt.savefig('Month of absence.png')


# In[160]:


sns.stripplot(x="Day of the week", y="Absenteeism time in hours", data=data, size = 10);
plt.savefig('Day of the week.png')


# In[161]:


sns.stripplot(x="Seasons", y="Absenteeism time in hours", data=data, size = 10);
plt.savefig('Seasons.png')


# In[165]:


ggplot(clean_data, aes(x='Transportation expense', y='Absenteeism time in hours')) +    geom_bar(fill= "Black") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Transportation expense") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[172]:


ggplot(clean_data, aes(x='Distance from Residence to Work', y='Absenteeism time in hours')) +    geom_bar(fill= "Green") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Distance from Residence to Work") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[168]:


sns.stripplot(x="Disciplinary failure", y="Absenteeism time in hours", data=data);
plt.savefig('Disciplinary failure.png')


# In[173]:


ggplot(clean_data, aes(x='Son', y='Absenteeism time in hours')) +    geom_bar(fill= "Purple") +    scale_color_brewer(type='diverging', palette=5) +    xlab("Son") + ylab("absenteeism in hr") + ggtitle("Absenteeism Analysis") + theme_bw()


# In[174]:


clean_data.columns

File: Employee Absenteeism.R
#lets clear the environment
rm(list = ls())


#set working directory 
setwd("S:/analytics/Projects/Employee Absenteeism")

getwd()

#load XLSX library
library("xlsx")

#Read the data

data = read.xlsx("Absenteeism_at_work_Project.xls", sheetIndex = 1)

#lets observe the dimension of data set
dim(data)

#lets get familiar with the column names
colnames(data)

#lets observe the structure, summary and class of data set respectively
str(data)
summary(data)
class(data)

#NOTE: as we can see all the 21 variables are in continuous form 

# NOTE: we have to categorise which of the attributes are continuous and which are categorical

### MISSING VALUE ANALYSIS ###
#lets check for the missing values
sum(is.na(data))  #number of missing values = 135\

#create a data frame with missing values 
missing_values = data.frame(apply(data,2,function(x)sum(is.na(x))))

names(missing_values)[1] = "Missing_Values"

missing_values$Variables = row.names(missing_values)

row.names(missing_values) = NULL


#Reaarange the columns
missing_values = missing_values[, c(2,1)]

#place the variables according to their number of missing values.
missing_values = missing_values[order(-missing_values$Missing_Values),]

#Calculate the missing value percentage
missing_values$percentage = (missing_values$Missing_Values/nrow(data) )* 100


#Store the missing value information in a csv file
write.csv(missing_values,"EA Project_Missing_value.csv", row.names = F)


### IMPORTANT NOTE: we will not drop any data because missing percentage is  less and data is also less as well. 


#created a copy of data for back up as we will need it further
df = data
#data = df
#imputing missing values

## create a missing value in any variable
data$Body.mass.index[12] #23
data$Body.mass.index[12] = NA

#Actual values = 23
#value from mean method = 26.68
#value from MEdian method = 25
#value from KNN ethod = 23

##lets apply mean method to impute this generated value and observe the result
#data$Body.mass.index[is.na(data$Body.mass.index)]  =  mean(data$Body.mass.index, na.rm = T) #26.68

##lets apply median method to impute this generated value and observe the result
#data$Body.mass.index[is.na(data$Body.mass.index)]  =  median(data$Body.mass.index, na.rm = T) #25


###Now convert the variables into their respective types
#structure of data
str(data)

data$ID = as.factor(as.character(data$ID))
data$Day.of.the.week = as.factor(as.character(data$Day.of.the.week))
data$Education = as.factor(as.character(data$Education))
data$Social.drinker = as.factor(as.character(data$Social.drinker))
data$Social.smoker = as.factor(as.character(data$Social.smoker))
data$Reason.for.absence = as.factor(as.character(data$Reason.for.absence))
data$Seasons = as.factor(as.character(data$Seasons))
data$Month.of.absence = as.factor(as.character(data$Month.of.absence))
data$Disciplinary.failure = as.factor(as.character(data$Disciplinary.failure))


#now apply KNN method to impute

library(DMwR)
data = knnImputation(data,k=3)

#after analysis of different data using median and KNN we find that KNN is more accurate and closerto original value than median for imputation

#Check presence of missing values once to confirm

apply(data,2, function(x){sum(is.na(x))})

#### NO more missing values are present here ###

#create subset of the dataset which have only numeric varaiables
#fisrt of all we will define a numeric index
numeric_index = sapply(data, is.numeric)
numeric_data = data[,numeric_index]

#now, we will store the numeric variables in a proper data frame which is a subset of dataset
numeric_data = as.data.frame(numeric_data)

#storing all the colnames excluding target variable name
cnames = colnames(numeric_data)[-12]

#outlier analysis
##box-plot distribution and outlier check
library(ggplot2)

for (i in 1:length(cnames)) {
  assign(paste0("gn",i), ggplot(aes_string( y = (cnames[i]), x= "Absenteeism.time.in.hours") , data = subset(data)) +
           stat_boxplot(geom = "errorbar" , width = 0.5) +
           geom_boxplot(outlier.color = "red", fill = "grey", outlier.shape = 20, outlier.size = 1, notch = FALSE)+
           theme(legend.position = "bottom")+
           labs(y = cnames[i], x= "Absenteeism.time.in.hours")+
           ggtitle(paste("Boxplot" , cnames[i])))
           #print(i)
}

options(warn = 0)

#lets plot the boxplots
gridExtra::grid.arrange(gn1, gn2,gn3, ncol=3)
gridExtra::grid.arrange(gn4,gn5,gn6, ncol=3)
gridExtra::grid.arrange(gn7,gn8,gn9, ncol =3)
gridExtra::grid.arrange(gn10,gn11, ncol =3 )

#here, either we can remove the outliers from all variables or can impute them by providin them value NA

#val = data$Distance.from.Residence.to.Work[data$Distance.from.Residence.to.Work %in% boxplot.stats(data$Distance.from.Residence.to.Work)$out]

#as the number of outliers is already less and volume of data is also low, we would go making each outlier NA and then will impute them

#getting outliers using boxplot.stat method
for (i in cnames) {
  print(i)
  val = data[,i][data[,i] %in% boxplot.stats(data[,i])$out]
  print(length(val))
  print(val)
}

#Make each outlier as NA
for (i in cnames) {
  val = data[,i][data[,i] %in% boxplot.stats(data[,i])$out]
  data[,i][data[,i] %in% val] = NA
}

#checking the missing values
sum(is.na(data)) #231

#find thenumber of missing values in each variable
for (i in cnames) {
  print(i)
  print(sum(is.na(data[,i])))
}

#Impute the values using KNN imputation method
data = knnImputation(data, k=3)

#Check again for missing value if present in case
sum(is.na(data)) #0


#lets confirm for the outliers by both the methods

for (i in 1:length(cnames)) {
  print(i)
  assign(paste0("g",i), ggplot(aes_string(y = (cnames[i]), x= "Absenteeism.time.in.hours"), data=subset(data))+
           stat_boxplot(geom = "errorbar", width= 0.5) +
           geom_boxplot(outlier.color = "red" , fill = "grey", outlier.shape = 18, outlier.size = 1, notch = FALSE) +
           theme(legend.position = "bottom")+
           labs(y = cnames[i], x="Absenteeism.time.in.hours") +
           ggtitle(paste("Box Plot of Employee data", cnames[i])))
  
}

gridExtra::grid.arrange(g1,g2, ncol = 2) 
gridExtra::grid.arrange(g3,g4,g5, ncol = 3)
gridExtra::grid.arrange(g6,g7,g8, ncol=3) 
gridExtra::grid.arrange(g9,g10,g11, ncol=3)

#no outlier detected

#Confirm using boxplot.stat method to see whether outlier exists
for (i in cnames) {
  print(i)
  val = data[,i][data[,i] %in% boxplot.stats(data[,i])$out]
  print(length(val))
  print(val)
}

#lets go Correlation plot
library(corrgram)
corrgram(na.omit(data))
dim(data)
corrgram(data[, cnames],order = F, upper.panel = panel.pie, text.panel = panel.txt, main = "correlation plot" )


#Here we can see in correlation plot that BMI and weight are highly positively correlated
#so we can drop either of the two variables bcz they are having  almost similar characteristics.


# Select the relevant numerical features

num_vars = c("Transportation.expense", "Distance.from.Residence.to.work", "Service.time", "Age", "Work.load.Average.day","Hit.target", "Son", "Pet", "Weight", "Height")
cnames
names(data)


#now, anova test regarding dimension reduction

install.packages("lsr")

library(lsr)

anova_test = aov(Absenteeism.time.in.hours ~ ID + Day.of.the.week + Education + Social.smoker + Social.drinker + Reason.for.absence + Seasons + Month.of.absence + Disciplinary.failure, data = data)
summary(anova_test)


#Dimension reduction
data_sorted = subset(data,select = -c(ID,Body.mass.index, Education, Social.smoker, Social.drinker, Seasons, Disciplinary.failure))


#feature selection

#copy of data set 
features_sorted = data_sorted

write.csv(data_sorted,"Sorted_Features.csv", row.names = F)


colnames(data_sorted)

#draw histogram on random variables to check if the distributions are normal

hist(data_sorted$Hit.target)
hist(data_sorted$Work.load.Average.day.)
hist(data_sorted$Weight)
hist(data_sorted$Transportation.expense)

#we will choose normalisation instead of standardisation bcz variables are not normally distributed

print(sapply(data_sorted ,is.numeric))
num_v = c("Transportation.expense", "Distance.from.Residence.to.Work","Service.time","Age","Work.load.Average.day.","Hit.target","Son","Pet","Weight","Height")
num_v

#feature scaling 
for (i in num_v) {
  print(i)
  data_sorted[,i] = ((data_sorted[,i] - min(data_sorted[,i])) /
                         (max(data_sorted[,i]) - min(data_sorted[,i])))
  
}

colnames(data_sorted)

#now we have normalized datastored in data_sorted
#Write this normalized data in csv form 
write.csv(data_sorted,"normalized_data_sorted.csv", row.names = F)

#######clean the environment except the data_sorted
library(DataCombine)
rmExcept(c("data_sorted","features_sorted"))


##as we have regression problem statement, so we will use decision trees for regression 
library("rpart")
library(MASS)
train_index = sample(1:nrow(data_sorted), 0.8* nrow(data_sorted))

train = data_sorted[train_index,]
test = data_sorted[-train_index,]

##model development (decision tree regression model)

#reg_model = rpart(Absenteeism.time.in.hours ~. , data = train, method = "anova")


#predicting reg_model for test cases
#predictions = predict(reg_model , test[,-14])

#RMSE OR MSEis the technique which can be used to evaluate the performance of a regression model

#here, i will use root mean square error technique to evaluate the performance of the model, moreover the data is a time series data

#library("DMwR")
#RMSE = regr.eval(test[,14], predictions, stats = 'rmse')
#RMSE #14.92

##accuracy = 85.02%
##  Thus in Decision tree regression model the error is 14.92 which tells that our model is 85.08% accurate____

#lets try with random forest regression model
###+____________Random Forest model+__
library("randomForest")
RF_model = randomForest(Absenteeism.time.in.hours~. , train, importance = TRUE,  ntree=100)

#Extract the rules generated as a result of random Forest model
library("inTrees")
rules_list = RF2List(RF_model)

#Extract rules from rules_list
rules = extractRules(rules_list, train[,-14])
rules[1:2,]


#Convert the rules in readable format
read_rules = presentRules(rules,colnames(train))
read_rules[1:2,]


#Determining the rule metric
rule_metric = getRuleMetric(rules, train[,-14], train$Absenteeism.time.in.hours)
rule_metric[1:2,]

#Prediction of the target variable data using the random Forest model
RF_prediction = predict(RF_model,test[,-14])
RMSE_RF = regr.eval(test[,13], RF_prediction, stats = 'rmse')
#RMSE = 7.88
#Accuracy = 92.12%

#Thus the error rate in Random Forest Model is 7.88% and the accuracy of the model is 100-7.88 = 92.12%.

###________________ LINEAR REGRESSION _______________________________

#library("usdm")
#LR_data_select = subset(data_sorted ,select = -c(Reason.for.absence,Day.of.the.week))
#colnames(LR_data_select)
#vif(LR_data_select[,-12])
#vifcor(LR_data_select[,-12], th=0.9)

####Execute the linear regression model over the data
#lr_model = lm(Absenteeism.time.in.hours~. , data = train)

#summary(lr_model)
#colnames(test)

###___Multiple R-squared:  0.2674,	Adjusted R-squared:  0.1953, which means our target variable can explain 26.74% of variance which is not acceptable.
#Predict the data 
#LR_predict_data = predict(lr_model, test[,1:13])

#Calculate MAPE
#MAPE(test[,14], LR_predict_data)
#library(Matrix)
#rmse(test[,14],LR_predict_data)

##linear regression model works best for continuous variables, but here in LR_data_select we hav
##________ Till here we have implemented Decision Tree, Random Forest and Linear Regression. Among all of these Random Forest is having highest accuracy.

#__ To calculate loss month wise we need to include month of absence variable again in our data set 
# LOSS = Work.load.average.per.day * Absenteeism.time.in.hours

data = features_sorted

colnames(data)
data$loss = data$Work.load.Average.day. * data$Absenteeism.time.in.hours

i=1


# NOW calculate Month wise loss encountered due to absenteeism of employees 

#Calculate loss in january
loss_jan = as.data.frame(data$loss[data$Month.of.absence %in% 1])
names(loss_jan)[1] = "Loss"
sum(loss_jan[1])
write.csv(loss_jan,"jan_loss.csv", row.names = F)

#Calculate loss in febreuary
loss_feb = as.data.frame(data$loss[data$Month.of.absence %in% 2])
names(loss_feb)[1] = "Loss"
sum(loss_feb[1])
write.csv(loss_feb,"feb_loss.csv", row.names = F)

#Calculate loss in march

loss_march = as.data.frame(data$loss[data$Month.of.absence %in% 3])
names(loss_march)[1] = "Loss"
sum(loss_march[1])
write.csv(loss_march,"march_loss.csv", row.names = F)

#Calculate loss in april

loss_apr = as.data.frame(data$loss[data$Month.of.absence %in% 4])
names(loss_apr)[1] = "Loss"
sum(loss_apr[1])
write.csv(loss_apr,"apr_loss.csv", row.names = F)


#calculate loss in may

loss_may = as.data.frame(data$loss[data$Month.of.absence %in% 5])
names(loss_may)[1] = "Loss"
sum(loss_may[1])
write.csv(loss_may,"may_loss.csv", row.names = F)


#calculate in june

loss_jun = as.data.frame(data$loss[data$Month.of.absence %in% 6])
names(loss_jun)[1] = "Loss"
sum(loss_jun[1])
write.csv(loss_jun,"jun_loss.csv", row.names = F)

#Calculate loss in july

loss_jul = as.data.frame(data$loss[data$Month.of.absence %in% 7])
names(loss_jul)[1] = "Loss"
sum(loss_jul[1])
write.csv(loss_jul,"jul_loss.csv", row.names = F)

#calculate loss in august

loss_aug = as.data.frame(data$loss[data$Month.of.absence %in% 8])
names(loss_aug)[1] = "Loss"
sum(loss_aug[1])
write.csv(loss_aug,"aug_loss.csv", row.names = F)

#Calculate loss in september

loss_sep = as.data.frame(data$loss[data$Month.of.absence %in% 9])
names(loss_sep)[1] = "Loss"
sum(loss_sep[1])
write.csv(loss_sep,"sep_loss.csv", row.names = F)

#calculate loss in october

loss_oct = as.data.frame(data$loss[data$Month.of.absence %in% 10])
names(loss_oct)[1] = "Loss"
sum(loss_oct[1])
write.csv(loss_oct,"oct_loss.csv", row.names = F)

#calculate loss in november

loss_nov = as.data.frame(data$loss[data$Month.of.absence %in% 11])
names(loss_nov)[1] = "Loss"
sum(loss_nov[1])
write.csv(loss_nov,"nov_loss.csv", row.names = F)

#calculate loss in december
loss_dec = as.data.frame(data$loss[data$Month.of.absence %in% 12])
names(loss_dec)[1] = "Loss"
sum(loss_dec[1])
write.csv(loss_dec,"dec_loss.csv", row.names = F)

